### PR TITLE
First Implementation of Spell feature

### PR DESCRIPTION
**Pull request by davidgoering**
Fairly straight forward.
Things to discuss:
- Do we want to be dependant on an external interface or do we want to keep a local json file as the underlying data?
- I wasn't quite sure how to manage a fuzzy search on the dictionary key. At the moment it will return the first spell it finds that matches (problem case ie "fireball" )

### CODE REVIEW

File: DiscordBot/SpellSearch.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            return MyStringBuilder.ToString();
+            
+
+        }
+
+        private Dictionary<string, Spell> map()
+        {
+            JsonSerializer serializer = new JsonSerializer();
+            using(var client = new WebClient()){
+                var json = client.DownloadString("https://raw.githubusercontent.com/tadzik/5e-spells/master/spells.json");
```
Comment (cameronpenner) [inline DiscordBot/SpellSearch.cs]: if we store the dictionary at class level, then we won't have to download it and convert it every time we want to check a spell. we could make it static :( and only build the dictionary if it hasn't been already. Or we could make SpellSearch a singleton :( or plug in a simple IoC container.

File: DiscordBot/SpellSearch.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            StringBuilder MyStringBuilder = new StringBuilder();
+            MyStringBuilder.AppendLine("> **Spell:** *" + result.Key + "*");
+            MyStringBuilder.AppendLine("> **Description:** " + result.Value.description);
+            MyStringBuilder.AppendLine("> **Casting Time:** " + result.Value.casting_time);
+            MyStringBuilder.AppendLine("> **Range:** " + result.Value.range);
+            MyStringBuilder.AppendLine("> **Level:** " + result.Value.level);
+            MyStringBuilder.AppendLine("> **School:** " + result.Value.school);
+            return MyStringBuilder.ToString();
+            
+
```
Comment (cameronpenner) [inline DiscordBot/SpellSearch.cs:22]: I'm pretty picky with my whitespace, we probably could get rid of line 16, 29, 30, and 40 in this file. Check out CodeMaid for visual studio, it's great for cleaning up simple formatting like this.

Comment (davidgoering) [inline DiscordBot/SpellSearch.cs:22]: Yeah like I would know the shortcut for auto formatting in IntelliJ in a heart beat... but thanks for the tip on the extension will install it right now ;)

File: DiscordBot/SpellSearch.cs (hunk 1 above)
Comment (davidgoering) [inline DiscordBot/SpellSearch.cs]: Good point, I will definitely move it out so that it is only done once.

File: DiscordBot/SpellSearch.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+{
+    public class SpellSearch
+    {
+
+        public string Search(string query)
+        {
+            Dictionary <string, Spell> spells = map();
+            KeyValuePair<string, Spell> result = spells.FirstOrDefault(kvp => kvp.Key.ToLower().Contains(query.ToLower()));
+            StringBuilder MyStringBuilder = new StringBuilder();
+            MyStringBuilder.AppendLine("> **Spell:** *" + result.Key + "*");
```
Comment (cameronpenner) [inline DiscordBot/SpellSearch.cs]: What does the "> " do here? will that be visible or is it a formatting thing?

Comment (davidgoering) [inline DiscordBot/SpellSearch.cs]: Well it was supposed to based on this: https://sourceforge.net/p/discord/wiki/markdown_syntax/#md_ex_bq

But it didn't work, so it shows up... but I think it looks better with the ">" than without... if we can figure out how to format it in a better way I am all ears.

File: DiscordBot/MonsterSearch.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    public class MonsterSearch
+    {
+        private static MonsterSearch instance;
+        List<Monster> monsters;
+
+        private MonsterSearch()
+        {
+        }
+
+        public static MonsterSearch Instance
```
Comment (cameronpenner) [inline DiscordBot/MonsterSearch.cs:19]: What do you think of making a singleton class and inheriting from that? less copy pasta code.

**Pull request merged**